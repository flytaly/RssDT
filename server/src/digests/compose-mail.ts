import { Enclosure, Feed, Item, ItemWithEnclosures, UserFeedWithOpts } from '#root/db/schema.js';
import { DateTime } from 'luxon';
import mjml2html from 'mjml';
import sanitizeHtml from 'sanitize-html';
import url from 'url';
import { TernaryState } from '../types/enums.js';
import { EnclosureWithTitle, Share } from '#root/types/index.js';
import { digestNames } from './digest-names.js';
import shareProviders from './share.js';
import themes, { HTMLMailTheme } from './themes.js';

function getUnsubscribeUrl(unsubscribeToken: string, id: string | number) {
  return `${process.env.FRONTEND_URL}/unsubscribe?token=${unsubscribeToken}&id=${id}`;
}

function ternaryToBool(value: TernaryState, defaultValue: boolean): boolean {
  if (value === TernaryState.default) return defaultValue;
  return value === TernaryState.enable;
}

const imagesTypes = [
  'image/gif',
  'image/jpg',
  'image/jpeg',
  'image/png',
  'image/svg+xml',
  'image/tiff',
  'image/webp',
];

/** Enclosures url could be too long. This function reduces them to filename and saves them as 'title' property. */
const addTitlesToEnclosures = (enclosures?: Enclosure[]) =>
  enclosures
    ? enclosures.reduce((acc, enc) => {
        if (!enc.url) return acc;
        const noSlashUrl = enc.url.endsWith('/') ? enc.url.slice(0, enc.url.length - 1) : enc.url;
        const { pathname } = url.parse(noSlashUrl);
        const filename = pathname?.split('/').pop();

        acc.push({ ...enc, title: filename || enc.url || '' });
        return acc;
      }, [] as EnclosureWithTitle[])
    : [];

const getImageFromEnclosures = (enclosures?: Enclosure[]) => {
  const enc = enclosures?.find(({ type }) => imagesTypes.includes(type || ''));
  return enc?.url || '';
};

export const composeHTML = (
  userFeed: UserFeedWithOpts,
  feed: Feed,
  items: Array<Item & { enclosures: Enclosure[] }>,
) => {
  const { user } = userFeed;
  const { options } = userFeed.user;
  const withToC = ternaryToBool(userFeed.withContentTable, options.withContentTableDefault);
  const withItemBody = ternaryToBool(userFeed.itemBody, options.itemBodyDefault);
  const withAttachments = ternaryToBool(userFeed.attachments, options.attachmentsDefault);
  const theme = themes[userFeed.theme as HTMLMailTheme];
  let resultStr = '<mjml>';
  resultStr += theme.header({
    title: feed.title || '',
    digestName: digestNames[userFeed.schedule],
  });
  if (withToC) {
    resultStr += theme.contentTable({ items });
  }
  resultStr += items.reduce((acc, item_) => {
    const { id, title, link } = item_;
    const imageUrl = item_.imageUrl || getImageFromEnclosures(item_.enclosures);
    const date = DateTime.fromJSDate(item_.pubdate || new Date())
      .setZone(user.timeZone)
      .setLocale(user.locale)
      .toLocaleString(DateTime.DATETIME_MED_WITH_SECONDS);
    const enclosures: EnclosureWithTitle[] = withAttachments
      ? addTitlesToEnclosures(item_.enclosures)
      : [];
    const share: Share[] =
      options.shareEnable && link
        ? shareProviders
            .filter((s) => !options.shareList?.length || options.shareList.includes(s.id))
            .map((s) => ({ ...s, url: s.getUrl(link, title || link) }))
        : [];
    const content = withItemBody ? item_.summary || item_.description : '';

    return (
      acc +
      theme.item({
        id,
        title: title || undefined,
        link: link || undefined,
        imageUrl,
        date,
        enclosures,
        content: content || undefined,
        share,
      })
    );
  }, '');

  resultStr += theme.footer({
    unsubscribeUrl: getUnsubscribeUrl(userFeed.unsubscribeToken, userFeed.id),
  });
  resultStr += '</mjml>';
  return mjml2html(resultStr);
};

export const composeText = (userFeed: UserFeedWithOpts, feed: Feed, items: Item[]) => {
  const { user } = userFeed;
  const { options } = userFeed.user;
  // const withToC = ternaryToBool(userFeed.withContentTable, options.withContentTableDefault);
  // const withAttachments = ternaryToBool(userFeed.attachments, options.attachmentsDefault);
  const withItemBody = ternaryToBool(userFeed.itemBody, options.itemBodyDefault);

  let result = '';

  // header
  result += `${feed.title} [${digestNames[userFeed.schedule]} digest]\n`;

  // items
  result += items.reduce((acc, item_) => {
    const date = DateTime.fromJSDate(item_.pubdate!)
      .setZone(user.timeZone)
      .setLocale(user.locale)
      .toLocaleString(DateTime.DATETIME_MED_WITH_SECONDS);
    const contentHtml = withItemBody ? item_.summary || item_.description || '' : '';
    const contentText = sanitizeHtml(contentHtml, { allowedTags: [] });
    return `${acc}\n\n${item_.title}\n${date}\n${item_.link}\n\n${contentText}\n`;
  }, '');

  // footer
  result += `\n\n========
You are receiving this digest because you subscribed to it on RssDT.com
\nUnsubscribe: ${getUnsubscribeUrl(userFeed.unsubscribeToken, userFeed.id)}`;

  return result;
};

export let composeDigest = (
  userFeed: UserFeedWithOpts,
  feed: Feed,
  items: ItemWithEnclosures[],
) => {
  if (!userFeed?.user?.options) {
    return { errors: [{ message: 'userFeed should have user and user.options objects' }] };
  }
  let html: string = '';
  let errors: Array<{ message: string }> | null = null;
  if (userFeed.theme === 'default') {
    ({ html, errors } = composeHTML(userFeed, feed, items!));
  }
  const text = composeText(userFeed, feed, items);
  return { text, html, errors };
};

export const composeDigestMock = (mock: typeof composeDigest) => {
  composeDigest = mock;
};
