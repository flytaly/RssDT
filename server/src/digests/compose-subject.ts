import { DigestSchedule } from '../types/enums.js';
import { digestNames } from './digest-names.js';

export const tokens = {
  title: '{{title}}',
  digestName: '{{digest}}',
};
const defaultTemplate = `${tokens.title}: ${tokens.digestName} digest`;

export const composeEmailSubject = (
  feedTitle: string,
  digestType: DigestSchedule,
  template?: string | null,
) => {
  const digestName = digestNames[digestType];

  if (!template) template = defaultTemplate;

  const replaceTokens = {
    [tokens.title]: feedTitle,
    [tokens.digestName]: digestName,
  };

  const replaceRegEx = new RegExp(Object.values(tokens).join('|'), 'gi');

  const subject = template.replace(replaceRegEx, (match) => replaceTokens[match]);

  return subject;
};
