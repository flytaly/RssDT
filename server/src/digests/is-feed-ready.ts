import { DateTime } from 'luxon';
// eslint-disable-next-line import/extensions
import { UserFeed } from '#entities';
import { scheduleHours, windowDuration } from '../constants.js';
import { DigestSchedule } from '../types/enums.js';

/**
 * Checks if the current userFeed is ready to be sent as a digest.
 * This function uses time relative to the user's time zone.
 *
 * A feed is "ready" when there weren't any emails sent (lastDigestSentAt)
 * past the most recent suitable hour. Suitable hours for "3 hourly digest"
 * are: 0, 3, 6, 9, 12, 15, 18, 21, 24. So if, for instance, current time is 15:15
 * and lastDigestSentAt=14:40 then the previous suitable hour is 15
 * and since 14:40 < 15:00 this userFeed is "ready".
 */

export let isFeedReady = (userFeed: UserFeed) => {
  const { lastDigestSentAt, schedule, user } = userFeed;
  if (schedule === DigestSchedule.disable) return false;
  if (schedule === DigestSchedule.realtime) return true;

  const zone = user.timeZone || 'GMT';

  const now = DateTime.local().setZone(zone);
  const lastDigestTime = DateTime.fromJSDate(lastDigestSentAt || new Date(0), { zone });
  const prevSuitableHour = now.hour - (now.hour % scheduleHours[schedule]);
  let prevSuitableTime = now //
    .set({ hour: prevSuitableHour, minute: 0, second: 0, millisecond: 0 });

  if (schedule === DigestSchedule.daily) {
    prevSuitableTime = prevSuitableTime.set({ hour: user.options.dailyDigestHour });
    const windowStart = prevSuitableTime;
    const windowEnd = windowStart.plus({ hours: windowDuration });
    if (now < windowStart || now > windowEnd) {
      return false;
    }
    if (prevSuitableTime > now) {
      prevSuitableTime = prevSuitableTime.minus({ days: 1 });
    }
  }

  if (lastDigestTime < prevSuitableTime) {
    return true;
  }

  return false;
};

export function isFeedReadyMock(mock: typeof isFeedReady) {
  isFeedReady = mock;
}
