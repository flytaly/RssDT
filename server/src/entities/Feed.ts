/* eslint-disable import/no-cycle */
import { Field, ObjectType } from 'type-graphql';
import {
    BaseEntity,
    Column,
    CreateDateColumn,
    Entity,
    OneToMany,
    PrimaryGeneratedColumn,
    UpdateDateColumn,
} from 'typeorm';
import { UserFeed } from './UserFeed';

@ObjectType()
@Entity()
export class Feed extends BaseEntity {
    @Field()
    @PrimaryGeneratedColumn()
    id!: number;

    @Field()
    @Column({ unique: true })
    url!: string;

    // TODO: should be protected
    @Field(() => [UserFeed], { nullable: true })
    @OneToMany(() => UserFeed, (userFeed) => userFeed.feed, { nullable: true })
    userFeeds: UserFeed[];

    @Field(() => Date)
    @CreateDateColumn()
    createdAt: Date;

    @Field(() => Date)
    @UpdateDateColumn()
    updatedAt: Date;
}
