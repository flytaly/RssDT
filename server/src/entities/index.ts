export * from './interfaces/index.js';
export * from './Feed.js';
export * from './Item.js';
export * from './User.js';
export * from './UsersToBeDeleted.js';
export * from './Options.js';
export * from './UserFeed.js';
export * from './Enclosure.js';
