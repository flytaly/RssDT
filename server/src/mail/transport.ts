import nodemailer from 'nodemailer';
import { IS_TEST } from '../constants';

export const transport = nodemailer.createTransport({
  port: +process.env.MAIL_PORT!,
  host: process.env.MAIL_SMTP,
  auth: {
    user: process.env.MAIL_USER,
    pass: process.env.MAIL_PASS,
  },
  ignoreTLS: IS_TEST,
});
