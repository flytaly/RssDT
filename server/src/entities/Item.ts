/* eslint-disable import/no-cycle */
import { Field, ObjectType } from 'type-graphql';
import {
    BaseEntity,
    Column,
    CreateDateColumn,
    Entity,
    ManyToOne,
    OneToMany,
    PrimaryGeneratedColumn,
} from 'typeorm';
import { Enclosure } from './Enclosure';
import { Feed } from './Feed';

@ObjectType()
@Entity()
export class Item extends BaseEntity {
    @Field()
    @PrimaryGeneratedColumn()
    id!: number;

    @Field({ nullable: true })
    @Column({ nullable: true })
    guid: string;

    @Field({ nullable: true })
    @Column({ nullable: true })
    pubdate: Date;

    @Field({ nullable: true })
    @Column({ nullable: true })
    link: string;

    @Field({ nullable: true })
    @Column({ type: 'text', default: '', nullable: true })
    title: string;

    @Field({ nullable: true })
    @Column({ type: 'text', default: '', nullable: true })
    description: string;

    @Field({ nullable: true })
    @Column({ type: 'text', default: '', nullable: true })
    summary: string;

    @Field({ nullable: true })
    @Column({ default: '', nullable: true })
    imageUrl: string;

    @Column({ nullable: true })
    feedId: number;

    @Field(() => Feed)
    @ManyToOne(() => Feed, (feed) => feed.items, { onDelete: 'CASCADE' })
    feed: Feed;

    @Field(() => [Enclosure], { nullable: true })
    @OneToMany(() => Enclosure, (enc) => enc.item, { nullable: true, eager: true, cascade: true })
    enclosures: Enclosure[];

    @Field(() => Date)
    @CreateDateColumn()
    createdAt: Date;
}
