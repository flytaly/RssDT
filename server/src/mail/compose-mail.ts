import { Feed } from '../entities/Feed';
import { Item } from '../entities/Item';
import { UserFeed } from '../entities/UserFeed';

export const composeDigest = (userFeed: UserFeed, feed: Feed, items: Item[]) => {
    // TODO:
    return { text: '', html: '', errors: [] };
};
