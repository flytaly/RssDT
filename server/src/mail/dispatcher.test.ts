import 'reflect-metadata';
import '../tests/test-utils/connection';
import { isFeedReady } from './is-feed-ready';
import { Feed } from '../entities/Feed';
import { generateItemEntity, generateUserWithFeed } from '../tests/test-utils/generate-feed';
import { User } from '../entities/User';
import { transport } from './transport';
import { composeDigest } from './compose-mail';
import {
    buildAndSendDigests,
    sendConfirmEmail,
    sendConfirmSubscription,
    sendPasswordReset,
} from './dispatcher';
import { UserFeed } from '../entities/UserFeed';
import { Item } from '../entities/Item';
import { composeEmailSubject } from './compose-subject';
import { DigestSchedule } from '../types/enums';

jest.mock('./is-feed-ready', () => ({
    isFeedReady: jest.fn((uf: UserFeed) => {
        expect(uf.schedule).toBeDefined();
        expect(uf.user.timeZone).toBeDefined();
        expect(uf.user.options.dailyDigestHour).toBeDefined();
        return true;
    }),
}));
jest.mock('./transport', () => ({ transport: { sendMail: jest.fn(async () => {}) } }));
jest.mock('./compose-mail', () => ({
    composeDigest: jest.fn((uf: UserFeed, feed: Feed, items: Item[]) => {
        expect(uf.schedule).toBeDefined();
        expect(uf.user.timeZone).toBeDefined();
        expect(uf.user.options.dailyDigestHour).toBeDefined();
        expect(feed.title).toBeDefined();
        expect(items.length > 0).toBeTruthy();
        expect(items[0].guid).toBeDefined();
        return { text: 'text', html: 'html', errors: [] };
    }),
}));

let user: User;
let feed: Feed;
let userFeed: UserFeed;

beforeAll(async () => {
    ({ user, feed, userFeed } = await generateUserWithFeed());
});

afterAll(async () => {
    await user.remove();
    await feed.remove();
});

const hour = 1000 * 60 * 60;

describe('Build and Send Digests', () => {
    const newItems: Item[] = [];
    const oldItems: Item[] = [];
    beforeAll(async () => {
        const now = Date.now();
        const lastDigestTime = now - hour * 24;
        userFeed.lastDigestSentAt = new Date(lastDigestTime);
        await userFeed.save();
        const item = (time: number) => generateItemEntity(feed.id, new Date(time));
        oldItems.push(await item(lastDigestTime - hour * 2));
        oldItems.push(await item(lastDigestTime - hour));
        oldItems.push(await item(lastDigestTime));
        newItems.push(await item(lastDigestTime + hour));
        newItems.push(await item(lastDigestTime + hour * 2));
        newItems.push(await item(lastDigestTime + hour * 3));
    });
    test('should get new items and send digest mail', async () => {
        await buildAndSendDigests(feed.id);
        expect(isFeedReady).toHaveBeenCalled();
        expect(composeDigest).toHaveBeenCalled();
        const call = (composeDigest as jest.Mock).mock.calls[0];
        const itemsPassed = call[2] as Item[];
        expect(itemsPassed).toHaveLength(newItems.length);
        const ids = new Set(newItems.map((i) => i.id));
        itemsPassed.forEach((item) => expect(ids.has(item.id)).toBeTruthy());

        expect(transport.sendMail).toHaveBeenCalledWith({
            from: process.env.MAIL_FROM,
            subject: composeEmailSubject(feed.title, userFeed.schedule, user.options.customSubject),
            to: user.email,
            html: 'html',
            text: 'text',
        });
    });

    test('should not send too old items even if they were created after last digest', async () => {
        (composeDigest as jest.Mock).mockClear();
        userFeed.lastDigestSentAt = new Date(0);
        await userFeed.save();
        // save very old item
        await generateItemEntity(feed.id, new Date(Date.now() - hour * 48));

        await buildAndSendDigests(feed.id);
        const call = (composeDigest as jest.Mock).mock.calls[0];
        const itemsPassed = call[2] as Item[];
        expect(itemsPassed.length).toBe(newItems.length + oldItems.length);

        const ids = new Set([...oldItems, ...newItems].map((i) => i.id));
        itemsPassed.forEach((item) => expect(ids.has(item.id)).toBeTruthy());
    });
});

describe('transaction emails', () => {
    const sendMail = transport.sendMail as jest.Mock;
    beforeEach(() => {
        sendMail.mockClear();
    });

    test('sendConfirmEmail', async () => {
        await sendConfirmEmail('email', 'token', 333);
        expect(sendMail.mock.calls).toMatchSnapshot();
    });
    test('sendPasswordReset', async () => {
        await sendPasswordReset('email', 'token', 333);
        expect(sendMail.mock.calls).toMatchSnapshot();
    });
    test('sendConfirmSubscription', async () => {
        await sendConfirmSubscription({
            email: 'email',
            token: 'token',
            userFeedId: 333,
            title: 'title',
            digestType: DigestSchedule.daily,
        });
        expect(sendMail.mock.calls).toMatchSnapshot();
    });
});
