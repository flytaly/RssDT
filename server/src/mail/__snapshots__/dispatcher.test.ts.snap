// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendConfirmEmail should  1`] = `
Array [
  Array [
    Object {
      "from": "TEST_MAIL_FROM",
      "html": "Please, confirm registration: <a href=\\"http://127.0.0.1/confirm-register?token=token&id=333\\">http://127.0.0.1/confirm-register?token=token&id=333</a><br><br>
        Ignore this message if you didn't register",
      "subject": "Confirm registration",
      "text": "Please, confirm registration: http://127.0.0.1/confirm-register?token=token&id=333.
        Ignore this message if you didn't register",
      "to": "email",
    },
  ],
]
`;

exports[`transaction emails sendConfirmEmail 1`] = `
Array [
  Array [
    Object {
      "from": "TEST_MAIL_FROM",
      "html": "Please, confirm registration: <a href=\\"http://127.0.0.1/confirm-register?token=token&id=333\\">http://127.0.0.1/confirm-register?token=token&id=333</a><br><br>
        Ignore this message if you didn't register",
      "subject": "Confirm registration",
      "text": "Please, confirm registration: http://127.0.0.1/confirm-register?token=token&id=333.
        Ignore this message if you didn't register",
      "to": "email",
    },
  ],
]
`;

exports[`transaction emails sendConfirmSubscription 1`] = `
Array [
  Array [
    Object {
      "from": "TEST_MAIL_FROM",
      "html": "Please, confirm subscription to <b>title</b> (Daily digest): <br>
        <a href=\\"http://127.0.0.1/confirm?token=token&id=333\\">http://127.0.0.1/confirm?token=token&id=333</a><br><br>
        Ignore this message if you didn't subscribe",
      "subject": "Confirm subscription to title",
      "text": "Please, confirm subscription to title (Daily digest):
        http://127.0.0.1/confirm?token=token&id=333

        Ignore this message if you didn't subscribe",
      "to": "email",
    },
  ],
]
`;

exports[`transaction emails sendPasswordReset 1`] = `
Array [
  Array [
    Object {
      "from": "TEST_MAIL_FROM",
      "html": "A password reset has been requested.<br>
        If it was you, visit this link to reset your password: <a href=\\"http://127.0.0.1/reset?token=token&id=333\\">http://127.0.0.1/reset?token=token&id=333</a><br><br>
        Otherwise, ignore this message.",
      "subject": "Reset password",
      "text": "A password reset has been requested.

If it was you, visit this link to reset your password: http://127.0.0.1/reset?token=token&id=333

Otherwise, ignore this message.",
      "to": "email",
    },
  ],
]
`;
